rtrlib/pfx/trie/trie-pfx.c

```diff
@@ -0,0 +1,472 @@
+                    pfx_table_notify_clients(pfx_table, &record, false);
+                }
+                rm_node = (trie_remove(root, &(root->prefix), root->len, 0));
+                assert(rm_node != NULL);
+                free(((struct node_data *) rm_node->data)->ary);
+                free(rm_node->data);
+                free(rm_node);
+            } while(rm_node != root);
+            if(i == 0)
+                pfx_table->ipv4 = NULL;
+            else
+                pfx_table->ipv6 = NULL;
+
+            pthread_rwlock_unlock(&(pfx_table->lock));
+        }
+    }
+    pthread_rwlock_destroy(&(pfx_table->lock));
+}
+
+int pfx_table_append_elem(struct node_data *data, const struct pfx_record *record)
+{
+    data->len++;
+    data->ary = realloc(data->ary, sizeof(struct data_elem) * data->len);
+    if(data->ary == NULL)
+        return PFX_ERROR;
+    data->ary[data->len - 1].asn = record->asn;
+    data->ary[data->len - 1].max_len = record->max_len;
+    data->ary[data->len - 1].socket = record->socket;
+    return PFX_SUCCESS;
+}
+
+int pfx_table_create_node(struct trie_node **node, const struct pfx_record *record)
+{
+    *node = malloc(sizeof(struct trie_node));
+    if(*node == NULL)
+        return PFX_ERROR;
+    (*node)->prefix = record->prefix;
+    (*node)->len = record->min_len;
+    (*node)->lchild = NULL;
+    (*node)->rchild = NULL;
+    (*node)->parent = NULL;
+
+    (*node)->data = malloc(sizeof(struct node_data));
+    if((*node)->data == NULL)
+        return PFX_ERROR;
+    ((struct node_data *) (*node)->data)->len = 0;
+    ((struct node_data *) (*node)->data)->ary = NULL;
+
+    return pfx_table_append_elem(((struct node_data *) (*node)->data), record);
+}
+
+struct data_elem *pfx_table_find_elem(const struct node_data *data, const struct pfx_record *record, unsigned int *index)
+{
+    for(unsigned int i = 0; i < data->len; i++) {
+        if(data->ary[i].asn == record->asn && data->ary[i].max_len == record->max_len && data->ary[i].socket == record->socket) {
+            if(index != NULL)
+                *index = i;
+            return &(data->ary[i]);
+        }
+    }
+    return NULL;
+}
+
+//returns pfx_table->ipv4 if record version is LRTR_IPV4 else pfx_table->ipv6
+inline struct trie_node *pfx_table_get_root(const struct pfx_table *pfx_table, const enum lrtr_ip_version ver)
+{
+    return ver == LRTR_IPV4 ? pfx_table->ipv4 : pfx_table->ipv6;
+}
+
+int pfx_table_del_elem(struct node_data *data, const unsigned int index)
+{
+    //if index is not the last elem in the list, move all other elems backwards in the array
+    if(index != data->len - 1) {
+        for(unsigned int i = index; i < data->len - 1; i++) {
+            data->ary[i] = data->ary[i+1];
+        }
+    }
+    data->len--;
+    data->ary = realloc(data->ary, sizeof(struct data_elem) * data->len);
+    if(data->len != 0 && data->ary == NULL)
```

Comment:
at a first glance I agree with @fho, that `free(&data->ary[data->len]);` is wrong.

Further I think the code should check if `len == 0` **before** doing the realloc, because according to man page:

> If size is zero and ptr is not NULL, a new,
     minimum sized object is allocated and the original object is freed.

Hence we need an explicit free here if empty